ellama.el

```diff
@@ -729,48 +730,170 @@ If EPHEMERAL non nil new session will not be associated with any file."
+  (with-slots (name) element
+    (format "```emacs-lisp\n(info \"%s\")\n```\n" name)))
+
+(cl-defmethod ellama-context-element-format
+  ((element ellama-context-element-info-node) (mode (eql 'org-mode)))
+  "Format the context ELEMENT for the major MODE."
+  (ignore mode)
+  (with-slots (name) element
+    (format "[[%s][%s]]"
+	    (replace-regexp-in-string
+	     "(\\(.?*\\)) \\(.*\\)" "info:\\1#\\2" name)
+	    (if (and ellama-chat-translation-enabled
+		     (not ellama--current-session))
+		(ellama--translate-string name)
+	      name))))
+
+;; Text context element
+
+(defclass ellama-context-element-text (ellama-context-element)
+  ((content :initarg :content :type string))
+  "A structure for holding information about a context element.")
+
+(cl-defmethod ellama-context-element-extract
+  ((element ellama-context-element-text))
+  "Extract the content of the context ELEMENT."
+  (oref element content))
+
+(cl-defmethod ellama-context-element-format
+  ((element ellama-context-element-text) (mode (eql 'markdown-mode)))
+  "Format the context ELEMENT for the major MODE."
+  (ignore mode)
+  (oref element content))
+
+(cl-defmethod ellama-context-element-format
+  ((element ellama-context-element-text) (mode (eql 'org-mode)))
+  "Format the context ELEMENT for the major MODE."
+  (ignore mode)
+  (oref element content))
+
 ;;;###autoload
 (defun ellama-context-add-file ()
   "Add file to context."
   (interactive)
-  (if-let* ((id ellama--current-session-id)
-	    (session (with-current-buffer (ellama-get-session-buffer id)
-		       ellama--current-session))
-	    (file-name (read-file-name "Select file: " nil nil t)))
-      (push (cons 'file file-name) (ellama-session-context session))
-    (push (cons 'file file-name) ellama--new-session-context)))
+  (let* ((file-name (read-file-name "Select file: " nil nil t))
+         (element (ellama-context-element-file :name file-name)))
+    (ellama-context-element-add element)))
 
 ;;;###autoload
 (defun ellama-context-add-buffer (buf)
   "Add BUF to context."
   (interactive "bSelect buffer: ")
-  (if-let* ((id ellama--current-session-id)
-	    (session (with-current-buffer (ellama-get-session-buffer id)
-		       ellama--current-session)))
-      (push (cons 'buffer buf) (ellama-session-context session))
-    (push (cons 'buffer buf) ellama--new-session-context)))
+  (let ((element (ellama-context-element-buffer :name buf)))
+    (ellama-context-element-add element)))
 
 ;;;###autoload
 (defun ellama-context-add-selection ()
   "Add file to context."
   (interactive)
-  (if-let* ((id ellama--current-session-id)
-	    (session (with-current-buffer (ellama-get-session-buffer id)
-		       ellama--current-session))
-	    ((region-active-p))
-	    (content (buffer-substring-no-properties (region-beginning) (region-end))))
-      (push (cons 'text content) (ellama-session-context session))
-    (push (cons 'text content) ellama--new-session-context)))
+  ;; TODO: Use region-active-p
```

Comment:
@s-kostyaev I added this TODO because I didn't understand what the `content` symbol is bound to in the else branch of the original code. Can you enlighten me?

Reply:
it should be when-let and without else branch. It's a bug.

Reply:
Or in else branch should be warning

Reply:
@r0man show warning to user in else branch. Something like "no active selection" or "no active region".

Reply:
@s-kostyaev Ok, I'm signaling a user error now with a `No active region` message.